src/com/facebook/buck/jvm/java/SourceReadingPackageFinder.java

```diff
@@ -0,0 +1,278 @@
+      int ch = reader.read();
+
+      switch (ch) {
+        case 'p':
+          return readPackage(reader);
+
+        case '/':
+          return readComment(reader);
+
+        default:
+          if (Character.isWhitespace(ch)) {
+            return readWhiteSpace(reader);
+          }
+          return new Token(Token.Kind.UNKNOWN, "");
+      }
+    }
+
+    private Token readWhiteSpace(Reader reader) throws IOException {
+      // Just keep going until we see something that's not whitespace.
+      for (;;) {
+        reader.mark(1);
+        int ch = reader.read();
+
+        if (ch == -1) {
+          return new Token(Token.Kind.UNKNOWN, "");
+        }
+
+        if (!Character.isWhitespace(ch)) {
+          reader.reset();
+          return new Token(Token.Kind.WHITE_SPACE, " ");
+        }
+      }
+    }
+
+    private Token readComment(Reader reader) throws IOException {
+      // We've already consumed the first '/'. Find out if we've got garbage, single or multi-line
+      int ch = reader.read();
+      switch (ch) {
+        case '/':
+          return readSingleLineComment(reader);
+
+        case '*':
+          return readMultiLineComment(reader);
+      }
+      return new Token(Token.Kind.UNKNOWN, "");
+    }
+
+    private Token readSingleLineComment(Reader reader) throws IOException {
+      // Read up until we see a line ending. Assume Mac, UNIX or DOS line endings.
+      StringBuilder value = new StringBuilder("//");
+
+      for (;;) {
+        int ch = reader.read();
+
+        switch (ch) {
+          case -1:
+            return new Token(Token.Kind.UNKNOWN, "");
+
+          // New-line always means the end of the line
+          case '\n':
+            return new Token(Token.Kind.COMMENT, value.toString());
+
+          // Carriage return might mean either a Mac EOL or a DOS one. Peek ahead
+          case '\r':
+            reader.mark(1);
+            ch = reader.read();
+            if (ch == -1) {
+              return new Token(Token.Kind.UNKNOWN, "");
+            }
+            if (ch == '\n') { // DOS newline
+              return new Token(Token.Kind.COMMENT, value.toString());
+            } else { // Mac newline
+              reader.reset();
+              return new Token(Token.Kind.COMMENT, value.toString());
+            }
+
+          default:
+            value.append((char) ch);
+            break;
+        }
+      }
+    }
+
+    private Token readMultiLineComment(Reader reader) throws IOException {
+      int last = reader.read();
+      if (last == -1) {
+        return new Token(Token.Kind.UNKNOWN, "");
+      }
+
+      StringBuilder value = new StringBuilder("/*");
+      value.append((char) last);
+      for (;;) {
+        int ch = reader.read();
+
+        if (ch == -1) {
+          return new Token(Token.Kind.UNKNOWN, "");
+        }
+
+        if (last == '*' && ch == '/') {
+          return new Token(Token.Kind.COMMENT, value.toString());
```

Comment:
I am not sure why the value is needed here, but since it is passed, it looks like this will contain the last `*`, but not the `/`.